### PR TITLE
first version of GITFans in Oscar

### PR DESCRIPTION
**Pull request by ThomasBreuer**
move the relevant code from the GITFans repository to the `experimental` directory of Oscar,
add the example in a test file,
leave out the outdated stuff

Since the rearrangements which I made today, the test example cannot be run because of a `std::bad_alloc` error in a call to `Polymake.polytope.equal_polyhedra`.
(Yesterday I saw this error for the first time, when I tried to run the example with trivial symmetry group, but the original example did not run into an error. Now this problem seems to be reproducible with the original test example.)

### CODE REVIEW

**Comment by fingolfin** [comment]
Perhaps @benlorenz or @lkastner can have with debugging the `std::bad_alloc` error?

**Comment by lkastner** [comment]
Yes, @ThomasBreuer wrote me yesterday and I did not get back to him yet. I will debug.

File: experimental/GITFans/GITFans.jl
Comment (fingolfin) [inline experimental/GITFans/GITFans.jl]:
```suggestion
julia> Oscar.GITFans.action_on_target(Q, G)
```

File: experimental/GITFans/GITFans.jl
Comment (fingolfin) [inline experimental/GITFans/GITFans.jl]:
```suggestion
Return orbit representatives of the group `G` on the set of those cones
whose defining rays are given by subsets `S` of the rows of the matrix `Q`,
such that the matrix `S` has full rank and such that the ideal `I` is
monomial-free (see [`is_monomial_free`](@ref)) w.r.t. the variables `x_i`
for which the `i`-th row of `Q` is not contained in `S`.
```
and similarly elsewhere

**Comment by fingolfin** [comment]
I don't see a `std::bad_alloc` error,  but there are still some doctest issues that would need to be adjusted

**Comment by fingolfin** [comment]
To clarify: for this to continue, I think @jankoboehm needs to be active, or anybody else who knows about GIT fans (suggestions welcome, perhaps @wdecker has some). But Thomas can't (and shouldn't) really push it forward on his own

**Comment by ThomasBreuer** [comment]
From my point of view, the general "show stopper" was the `std::bad_alloc` error. As far as I understand, this problem has already been fixed in Polymake, and as soon as this fix is available in Oscar, we can merge this pull request.
I will change a few ugly hacks in the code where matrix groups in characteristic zero are involved; these hacks are not necessary anymore.
Concerning the rest of the code and its further development, I agree with Max that the experts are asked.

File: experimental/GITFans/GITFans.jl
Comment (fingolfin) [inline experimental/GITFans/GITFans.jl:133]: just to  clarify, is this simply meant to trigger some computations whose results are ignored (for now)? Or what is the motivation?

File: experimental/GITFans/GITFans.jl
Comment (fingolfin) [inline experimental/GITFans/GITFans.jl]: This gives an error now. I guess that means my change suggestion (which was applied) was bad and we need to add the `Oscar.` back...

File: experimental/GITFans/GITFans.jl
Comment (fingolfin) [inline experimental/GITFans/GITFans.jl]: Move this to the bib file?

File: experimental/GITFans/GITFans.jl
Comment (lkastner) [reply experimental/GITFans/GITFans.jl:133]: The results are only ignored at toplevel. polymake uses these results internally for checking equality of the polyhedra. The problem is that polymake creates a giant LP otherwise. In the next polymake release this is fixed and these lines can go away. I even mentioned this in the commit message to it is easy to find out what to remove once that happens.

File: experimental/GITFans/GITFans.jl
Comment (fingolfin) [inline experimental/GITFans/GITFans.jl]:
```suggestion
    return BitSet(perm[i] for i in bitset)
```
Or replace `perm` by a `PermGroupElem` and then just write `bitset^perm`. :-)

File: experimental/GITFans/GITFans.jl
Comment (fingolfin) [inline experimental/GITFans/GITFans.jl]:
```suggestion
end
```
Several similar instances of redundant semicolons should also be updated

File: experimental/GITFans/GITFans.jl
Comment (fingolfin) [inline experimental/GITFans/GITFans.jl]:
```suggestion
    return [bitlist_oper(bitset_tuple[i], perm_tuple[i]) for 1:length(bitset_tuple)]
```

File: experimental/GITFans/GITFans.jl
Comment (fingolfin) [inline experimental/GITFans/GITFans.jl:43]: Perhaps this works, too?
```suggestion
    return minimum(current_orbit)
```

File: experimental/GITFans/GITFans.jl
Comment (fingolfin) [inline experimental/GITFans/GITFans.jl]: This can be transformed first to
```suggestion
        for i in [Vector{Int64}(orb[1]) for orb in orbs]
```
and then to the following, which avoids creating a temporary `Vector`
```suggestion
        for orb in orbs
          i = Vector{Int64}(orb[1])
```

File: experimental/GITFans/GITFans.jl
Comment (fingolfin) [reply experimental/GITFans/GITFans.jl:133]: Thanks for the clarification!

Comment (fingolfin) [reply experimental/GITFans/GITFans.jl:133]: Of course now I am also wondering why this code directly calls `Polymake.polytope.Cone` instead of using an `Oscar.Cone` object (answer is probably: because this code predates `Oscar.Cone` ;-) but then perhaps it should be brought forward to use that? Alternatively, perhaps a TODO comment to that effect ought to be inserted?

File: experimental/GITFans/GITFans.jl
Comment (fingolfin) [inline experimental/GITFans/GITFans.jl]:
```suggestion
    result_cones = [hash_to_cone(orbit_list, x) for x in hash_list]
```

File: experimental/GITFans/GITFans.jl
Comment (fingolfin) [inline experimental/GITFans/GITFans.jl]: What is the `collect(Set(...))` bit about? Is this to get a duplicate free list? Then perhaps `unique` could be used instead? And then `sort! instead of `sort` to avoid yet another allocation?
```suggestion
    allrays = sort!(unique(vcat(rays_maxcones...)))
```

File: experimental/GITFans/GITFans.jl
Comment (fingolfin) [inline experimental/GITFans/GITFans.jl]:
```suggestion
    poly_list = [evaluate(f, vars_to_zero, fill(R(0), length(vars_to_zero))) for f in gens(SingI)]
```

File: experimental/GITFans/GITFans.jl
Comment (fingolfin) [inline experimental/GITFans/GITFans.jl]:
```suggestion
    mats_transp = [Matrix{Rational{BigInt}}(x.X)' for x in matgens]
```

File: experimental/GITFans/GITFans.jl
Comment (fingolfin) [inline experimental/GITFans/GITFans.jl]: Julia best practice is to write `MODULE.@MACRO` not `@MODULE.MACRO`
```suggestion
    pm_gens = Polymake.@convert_to Array{Array{Int}} rewr2
```

File: experimental/GITFans/GITFans.jl
Comment (fingolfin) [inline experimental/GITFans/GITFans.jl]:
```suggestion
    pm_mc_reps = Polymake.@convert_to Array{Array{Int}} index_result_cones
```

**Review by fingolfin — COMMENTED** [review]
Left some remarks... you may not wish to address all of them now, but perhaps we can at least add some notes to remind about them?

All in all, this code is still quite some "un-julia-ish".

**Comment by ThomasBreuer** [comment]
@fingolfin Thanks for the comments. I will address most of them now.

File: experimental/GITFans/GITFans.jl
Comment (ThomasBreuer) [reply experimental/GITFans/GITFans.jl]: Yes.

File: experimental/GITFans/GITFans.jl
Comment (ThomasBreuer) [reply experimental/GITFans/GITFans.jl:43]: We want the smallest entry w.r.t. `leq`.
I did not find a way to compute `minimum` for an array w.r.t. a prescribed comparison function.

**Review by fingolfin — APPROVED** [review]
Let's merge this now (assuming the tests still pass -- I just had to resolve a merge conflict caused by another PR being merged) and improve it further later. Thanks!

**Pull request merged**
